src/platform/auth.ts

```diff
@@ -312,10 +313,14 @@ export function getUser(
 	});
 }
 
-export const findUser = (
+export const findUser = <
+	T extends Pick<DbUser, TProps>,
+	TProps extends keyof DbUser
+>(
 	loginInfo: string,
-	tx?: Tx,
-): Promise<AnyObject | undefined> => {
+	tx: Tx | undefined = undefined,
+	$select: TProps[] = ['id', 'actor', 'username', 'password'] as TProps[],
```

Comment:
The ` as TProps[]` should be unnecessary:
```suggestion
	$select: TProps[] = ['id', 'actor', 'username', 'password'],
```

Reply:
Interestingly enough, having `const defaultFindUserProps: Array<keyof DbUser> = ['id', 'actor', 'username', 'password'];` right above would work, but TS errors w/o this.
Doing `findUser('thgreasi', undefined, ['id', 'actor', 'username', 'password'])` works fine.
That's also how we ended up typing such functions in the cloud API as well.

Reply:
So the current typing can allow for an invalid return type, there's 2 ways of fixing this:
1. remove the $select parameter altogether since it wasn't there before and probably isn't necessary:
```ts
export const findUser = (
	loginInfo: string,
	tx: Tx | undefined = undefined,
): Promise<Pick<DbUser, typeof $select[number]> | undefined> => {
	if (!loginInfo) {
		return Promise.resolve(undefined);
	}
	const $select = ['id' as const, 'actor' as const, 'username' as const, 'password' as const];

	let loginField;
	if (loginInfo.includes('@')) {
		loginField = 'email';
	} else {
		loginField = 'username';
	}

	return resinApi
		.get({
			resource: 'user',
			passthrough: {
				req: root,
				tx,
			},
			options: {
				$filter: {
					$eq: [
						{
							$tolower: { $: loginField },
						},
						{
							$tolower: loginInfo,
						},
					],
				},
				$select,
			},
		})
		.then(([user]: [Pick<DbUser, typeof $select[number]>?]) => user);
};
```
2. use an overload
```ts
export function findUser(
	loginInfo: string,
	tx?: Tx | undefined,
	$select?: undefined,
): Promise<Pick<DbUser, 'username' | 'password' | 'id' | 'actor'> | undefined>;
export function findUser<
	T extends Pick<DbUser, TProps>,
	TProps extends keyof DbUser
>(
	loginInfo: string,
	tx: Tx | undefined,
	$select: TProps[],
): Promise<T | undefined>;
export function findUser(
	loginInfo: string,
	tx: Tx | undefined = undefined,
	$select = ['id' as const, 'actor' as const, 'username' as const, 'password' as const],
): Promise<Pick<DbUser, typeof $select[number]> | undefined> {
	if (!loginInfo) {
		return Promise.resolve(undefined);
	}

	let loginField;
	if (loginInfo.includes('@')) {
		loginField = 'email';
	} else {
		loginField = 'username';
	}

	return resinApi
		.get({
			resource: 'user',
			passthrough: {
				req: root,
				tx,
			},
			options: {
				$filter: {
					$eq: [
						{
							$tolower: { $: loginField },
						},
						{
							$tolower: loginInfo,
						},
					],
				},
				$select,
			},
		})
		.then(([user]: [Pick<DbUser, typeof $select[number]>?]) => user);
};
```

As a note I'd prefer `['id', 'actor', 'username', 'password'] as const` to `['id' as const, 'actor' as const, 'username' as const, 'password' as const]` but the trailing `as const` makes it readonly and fails to compile..

Reply:
How about using `['id', 'actor', 'username', 'password'] as const` along with an extra `Writable` type modifier as in the initial commit, so that we can avoid repeating `as const` everywhere?
See: https://github.com/balena-io/open-balena-api/pull/160/commits/d2289415a935497d236123dde5e29250cdfef144

Reply:
Yeah, I even searched for a type that was removing readonly as I'd sworn I'd seen it in use before! I guess now I know why I thought that.. :P